### DISCUSSION TITLE
How to do auto complete from database table

### DISCUSSION THREAD

**Discussion by heejit** [sqlpage/SQLPage#220] [General]
First SQLPage  is very nice many thanks for sharing.

How do I allow user to select / auto complete a input box from sql table which contain 10K rows (item_names) user need to input few letter to start search.

**Comment by lovasoa** [comment]
Hello! There is no such component in SQLPage yet... Would you be interested in contributing one? I can help !

**Reply by heejit** [reply]
Yes if you can guide me I can try.

**Comment by lovasoa** [comment]
What I would do is reuse the "tags input" element from the tabler css library that SQLPage embeds:

![image](https://github.com/lovasoa/SQLpage/assets/552629/924af4d6-7188-4ffa-b4ac-e420def88c8e)

```html
<div class="mb-3">
  <label class="form-label">Tags input</label>
  <select type="text" class="form-select tomselected ts-hidden-accessible" placeholder="Select tags" id="select-tags" value="" multiple="multiple" tabindex="-1">
    <option value="HTML">HTML</option>
    <option value="JavaScript">JavaScript</option>
    <option value="CSS">CSS</option>
    <option value="jQuery">jQuery</option>
    <option value="Bootstrap">Bootstrap</option>
    <option value="Ruby">Ruby</option>
    <option value="Python">Python</option>
  </select><div class="ts-wrapper form-select multi"><div class="ts-control"><input tabindex="0" role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-controls="select-tags-ts-dropdown" id="select-tags-ts-control" placeholder="Select tags" type="select-multiple"></div></div>
</div>
```

https://tabler.io/preview

You can [follow the instructions to create your own custom component](https://sql.ophir.dev/custom_components.sql), or contribute directly to the [form component](https://github.com/lovasoa/SQLpage/blob/main/sqlpage/templates/form.handlebars)

The way I would do it is adding a new row level parameter called `options_source` that the user would set to the name of a sql file to call to get the list of options.

Then the user could do 
### `fom.sql`
```sql
select 'form' as component;
select 'My choice' as name, 'my_items.sql' as options_source;
```
### `my_items.sql`
```sql
SELECT 'json' AS component, (
    SELECT JSON_GROUP_ARRAY(item_name) FROM items WHERE item_name LIKE '%' || $search
) AS contents;
```

what do you think ?

**Reply by heejit** [reply]
I am beginner in web programming. 

I don't understand the Sqlpage inner working.

I have little experience with Go and Htmx

I have small app written in Go and htmx.org.

What I did is when user press enter key on a input-box a javascript function send ajax request to 
backend and in response get a dialog which contain a Html Table and 1st column of each row is a button which call other javascript function to handle the user selection and closing of dialog. 

What I understand as per your suggestion is on every input change sqlpage will call my_item.sql to get a new item list and update the select component right ?

**Reply by lovasoa** [reply]
Yes, exactly ! I'm turning this discussion into a feature request, so that we can keep track of it :)

**Reply by lovasoa** [reply]
https://github.com/lovasoa/SQLpage/issues/221

**Comment by lovasoa** [comment]
Support for autocomplete has been added and will be released in 0.20.3 :tada: 

You will be able to use a `select` input form with a `dropdown` property and create a beautiful input with autocomplete. 

https://sql.ophir.dev/documentation.sql?component=form#component
